march_hardware_interface/src/march_hardware_interface_node.cpp

```diff
@@ -45,16 +45,22 @@ int main(int argc, char** argv)
       ros::Duration(march.getEthercatCycleTime() * ros::param::param<int>("~control_loop_multiplier", 1) / 1000.0));
 
   controller_manager::ControllerManager controller_manager(&march, nh);
+  ros::Time last_update_time = ros::Time::now() - rate.expectedCycleTime();
 
   while (ros::ok())
   {
     const ros::Time now = ros::Time::now();
     try
     {
-      march.read(now, rate.expectedCycleTime());
-      march.validate();
-      controller_manager.update(now, rate.expectedCycleTime());
-      march.write(now, rate.expectedCycleTime());
+      if (march.getTrainReturned())
+      {
+        march.setTrainReturned(false);
```

Comment:
Essentially what you want to do here is wait for new data to arrive. You can probably do this a lot more efficient using [`std::condition_variable`](https://en.cppreference.com/w/cpp/thread/condition_variable), so that you can be notified when a new train arrived. However, when implementing this I am wondering what the advantage then is of having the ethercat train is separate thread anyway, since now you are forcing it to be sequential, which you can do a lot easier using a single thread. What is the core reason that you want to implement this? What problem does it solve?

Reply:
I think that ideally you want to update your controller as fast as there is new information. Although it could be argued how much it would matter.

Reply:
You want to prevent the controller updating before there is a data update, because then it will think that the velocity is zero (and the velocity of the next timestep will be 2x the normal). I thought about putting them in the same thread, but I wanted to avoid the controller somehow gobbeling up computation and hence delaying the train. A signal using std::condition_variable seems like a good idea but little stupid me did not know how to do this. I will look at it tomorrow :)

Reply:
I implemented something but I am far from certain it will work :sweat_smile: Please have a look and bestow me your wisdom :)